src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotAddOrRemoveVirtualModifierTests.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Xunit;
+
+namespace Microsoft.DotNet.ApiCompatibility.Tests
+{
+    public class CannotAddOrRemoveVirtualModifierTests
+    {
+        [Fact]
+        public static void DifferencesReported()
+        {
+            string leftSyntax = @"
+namespace CompatTests
+{
+  public abstract class First {
+    public virtual void F() {}
+    public void G() {}
+    public abstract void H();
+  }
+}
+";
+            string rightSyntax = @"
+namespace CompatTests
+{
+  public abstract class First {
+    public void F() {}
+    public virtual void G() {}
+    public virtual void H() {}
+  }
+}
+";
+            IAssemblySymbol left = SymbolFactory.GetAssemblyFromSyntax(leftSyntax);
+            IAssemblySymbol right = SymbolFactory.GetAssemblyFromSyntax(rightSyntax);
+            ApiComparer differ = new();
+            IEnumerable<CompatDifference> differences = differ.GetDifferences(new[] { left }, new[] { right });
+            CompatDifference[] expected = new[]
+            {
+                new CompatDifference(DiagnosticIds.CannotRemoveVirtualModifier, string.Empty, DifferenceType.Removed, "M:CompatTests.First.F"),
+                new CompatDifference(DiagnosticIds.CannotAddVirtualModifier, string.Empty, DifferenceType.Added, "M:CompatTests.First.G"),
+            };
+            Assert.Equal(expected, differences);
+        }
+
+        [Fact]
+        public static void Null()
+        {
+            string leftSyntax = @"
+namespace CompatTests
+{
+  public class First {
+    public virtual void F() {}
+  }
+}
+";
+            string rightSyntax = @"
+namespace CompatTests
+{
+  public class First {
+    public virtual void G() {}
+  }
+}
+";
+            IAssemblySymbol left = SymbolFactory.GetAssemblyFromSyntax(leftSyntax);
+            IAssemblySymbol right = SymbolFactory.GetAssemblyFromSyntax(rightSyntax);
+            ApiComparer differ = new();
+            IEnumerable<CompatDifference> differences = differ.GetDifferences(new[] { left }, new[] { right });
+            Assert.NotEmpty(differences);
```

Comment:
I don't think this test is testing anything really valuable. What you want to make sure here, is to test that the differences reported do NOT contain the diagnostics that are thrown from your new rule, and instead only MembersMustExist diagnostics.

Reply:
My initial goal was to make sure that null members on the left or right do not cause the test to fail with a `NullReferenceException`.

Reply:
right, and I think that is fine, but this Assert is not doing that. The fact that an exception is not thrown is simply by calling `differ.GetDifferences` and having that not throw an exception. What I'm saying, is that it would be good to also ensure not only that the code didn't hit any exceptions, but also that the rule didn't emit any diagnostics incorrectly.